Implement VID Verification in OpCreds cluster (2/2)

tcarmelveilleux opened the pull request:
#### Changes

- Add VVS/VVSC storage in PersistentStorageOpCertStore
- Finished all attributes and commands in Opcreds cluster
- Bumped Opcreds revision to 2
- Added all necessary testability features to matter_testing support
  for list subscriptions

#### Testing

- Integration tests through TC-OPCREDS-3.9
- More unit tests to follow, but coverage through TC-OPCREDS-3.9 is near
  exhaustive.

tcarmelveilleux committed:
Implement VID Verification in OpCreds cluster (2/2)

- Add VVS/VVSC storage in PersistentStorageOpCertStore
- Finished all attributes and commands in Opcreds cluster
- Bumped Opcreds revision to 2
- Added all necessary testability features to matter_testing support
  for list subscriptions
- NOTE: this includes #38445 temporarily until merged

Testing done:

- Integration tests through TC-OPCREDS-3.9
- More unit tests to follow, but coverage through TC-OPCREDS-3.9 is near
  exhaustive.

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into add-set-vid-verify-statement

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into add-set-vid-verify-statement

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into add-set-vid-verify-statement

tcarmelveilleux committed:
Merge remote-tracking branch 'origin/add-set-vid-verify-statement' into add-set-vid-verify-statement

tcarmelveilleux committed:
Fix TC-CGEN-2.9

- Fabric table was incorrectly read, which could lead to issues
  when the fabric table has more fields than before.

cecille commented on src/credentials/PersistentStorageOpCertStore.cpp:
```diff
@@ -494,5 +712,59 @@ CHIP_ERROR PersistentStorageOpCertStore::GetCertificate(FabricIndex fabricIndex,
+
+    uint8_t storageBuffer[kMaxCHIPCertLength];
+    uint16_t keySize = sizeof(storageBuffer);
+    static_assert(kMaxCHIPCertLength > (2 * (Crypto::kVendorIdVerificationStatementV1Size)),
```
Does this have alternate SDK impact as well? ie, should this be something that's specifically called out in the spec?

tcarmelveilleux replied:
It doesn't really. This is cheap A-B-A issues prevention attempt.

tcarmelveilleux commented on src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:
```diff
@@ -247,10 +247,13 @@ OperationalCredentialsAttrAccess gAttrAccess;
 CHIP_ERROR OperationalCredentialsAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
 {
-    VerifyOrDie(aPath.mClusterId == Clusters::OperationalCredentials::Id);
+    static constexpr uint16_t kCurrentClusterRevision = 2u;
```
Done

tcarmelveilleux commented on src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:
```diff
@@ -1232,8 +1235,58 @@ bool emberAfOperationalCredentialsClusterSetVIDVerificationStatementCallback(
+        return true;
+    }
+
+    if (commandData.vendorID.HasValue() && !IsVendorIdValidOperationally(static_cast<VendorId>(commandData.vendorID.Value())))
```
Done, good catch.

tcarmelveilleux commented on src/credentials/FabricTable.cpp:
```diff
@@ -2246,4 +2240,63 @@ CHIP_ERROR FabricTable::SignVIDVerificationRequest(FabricIndex fabricIndex, Byte
+    FabricInfo * fabricInfo = GetMutableFabricByIndex(fabricIndex);
+    VerifyOrReturnError(fabricInfo != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    bool isTargetFabricPending = (GetPendingNewFabricIndex() == fabricIndex) ||
```
Done

tcarmelveilleux commented on src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:
```diff
@@ -1232,8 +1235,58 @@ bool emberAfOperationalCredentialsClusterSetVIDVerificationStatementCallback(
+    // Handle dirty-marking if anything changed.
+    if (fabricChangesOccurred)
+    {
+        MatterReportingAttributeChangeCallback(commandPath.mEndpointId, OperationalCredentials::Id,
```
`NOCs` is `C` quality and not reportable. Will add a  comment.

tcarmelveilleux commented on src/controller/python/chip/ChipDeviceCtrl.py:
```diff
@@ -1080,12 +1080,12 @@ def GetFabricIdInternal(self):
 
         Returns:
-            int: The compressed fabric ID as a 64-bit integer.
+            int: fabric index of applying to the controller (fabricIndex in local fabric table does not match remote)
```
Fixed the comment.

tcarmelveilleux commented on src/credentials/FabricTable.cpp:
```diff
@@ -2246,4 +2240,63 @@ CHIP_ERROR FabricTable::SignVIDVerificationRequest(FabricIndex fabricIndex, Byte
+        if (static_cast<uint16_t>(fabricInfo->GetVendorId()) != vendorId.Value())
+        {
+            fabricInfo->SetVendorId(static_cast<VendorId>(vendorId.Value()));
+            if (!isTargetFabricPending)
```
Updated comment.

tcarmelveilleux commented on src/credentials/PersistentStorageOpCertStore.cpp:
```diff
@@ -279,6 +338,8 @@ CHIP_ERROR PersistentStorageOpCertStore::UpdateOpCertsForFabric(FabricIndex fabr
     VerifyOrReturnError(!noc.empty() && (noc.size() <= Credentials::kMaxCHIPCertLength), CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrReturnError(icac.size() <= Credentials::kMaxCHIPCertLength, CHIP_ERROR_INVALID_ARGUMENT);
+    // Can't set an ICAC if we have installed a VVSC, until the VVSC is gone.
+    VerifyOrReturnError(!HasVvscForFabric(fabricIndex), CHIP_ERROR_INCORRECT_STATE);
```
> Does that imply that you need to send a SetVIDVerficationStatement command to the device before updating the NOC if there's a VVSC?

Yes

Will add a log.

tcarmelveilleux committed:
Address review comments from cecille@

restyled-commits committed:
Restyled by clang-format

cecille commented on src/python_testing/TC_OPCREDS_3_9.py:
```diff
@@ -161,194 +168,490 @@ def get_unassigned_fabric_index(fabric_indices: list[int]) -> int:
+            asserts.assert_equal(cr2_fabrics_entry.vendorID, vendorID,
+                                 "Did not get the expected value set for VendorID field of Fabrics list for TH2's fabric.")
+
+        with test_step(description="Invoke SetVIDVerificationStatement with maximum-sized VVSC on TH1's fabric, outside fail-safe. Verify INVALID_COMMAND due to presence of ICAC."):
```
As far as I can tell, the command description doesn't require a failsafe, does it? Does the behaviour change when the failsafe is armed if there wasn't a prior add/update noc call?

cecille reviewed (commented):
if this is for OPCREDS-3.8, does the test do this part? * Ensure that only accessing fabric's VidVerificationStatement can be set/clear

cecille reviewed (approved):
Follow ups on test comments are OK. Expect maybe for clarifying the test number.

tcarmelveilleux commented on src/python_testing/TC_OPCREDS_3_9.py:
```diff
@@ -161,194 +168,490 @@ def get_unassigned_fabric_index(fabric_indices: list[int]) -> int:
+def get_entry_for_fabric(fabric_index: int, entries: list[object]) -> object:
+    for entry in entries:
+        if not hasattr(entry, "fabricIndex"):
+            return None
```
All entries need to have fabricIndex, it's baked in and mandatory for fabric scoped lists.

tcarmelveilleux commented on src/python_testing/TC_OPCREDS_3_9.py:
```diff
@@ -161,194 +168,490 @@ def get_unassigned_fabric_index(fabric_indices: list[int]) -> int:
-            await self.send_single_cmd(cmd=opcreds.Commands.SignVIDVerificationRequest(fabricIndex=cr2_fabric_index, clientChallenge=CHALLENGE_TOO_BIG))
+    @async_test_body
+    async def test_TC_OPCREDS_3_9(self):
+        # TODO(test_plans#5046): actually make the test follow final test plan. For now
```
It's still a TODO as test plan being built in parallel.

tcarmelveilleux commented on src/python_testing/TC_OPCREDS_3_9.py:
```diff
@@ -161,194 +168,490 @@ def get_unassigned_fabric_index(fabric_indices: list[int]) -> int:
-        with asserts.assert_raises(InteractionModelError) as exception_context:
-            await self.send_single_cmd(cmd=opcreds.Commands.SignVIDVerificationRequest(fabricIndex=cr2_fabric_index, clientChallenge=CHALLENGE_TOO_BIG))
+    @async_test_body
+    async def test_TC_OPCREDS_3_9(self):
```
This is opcreds 3_8. I will rename it now since it's causing confusion.

tcarmelveilleux replied:
Renamed.

tcarmelveilleux committed:
Renamed TC-OPCREDS-3.9 to TC-OPCREDS-3.8

tcarmelveilleux committed:
Added comments to AttributeMatcher

tcarmelveilleux committed:
Renamed test in TC_OPCREDS_3_8.py

restyled-commits committed:
Restyled by autopep8

tcarmelveilleux commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -291,6 +291,30 @@ class AttributeValue:
     timestamp_utc: Optional[datetime] = None
 
 
+class AttributeMatcher:
```
Done

tcarmelveilleux committed:
More logging in TC-CGEN-2.9

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into add-set-vid-verify-statement

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into add-set-vid-verify-statement

tcarmelveilleux committed:
Add a verification against some node ID to diagnose CI

restyled-commits committed:
Restyled by clang-format

tcarmelveilleux commented on src/credentials/FabricTable.cpp:
```diff
@@ -118,6 +119,7 @@ CHIP_ERROR FabricInfo::Init(const FabricInfo::InitParams & initParams)
     Reset();
 
     mNodeId                  = initParams.nodeId;
+    VerifyOrDie(mNodeId != kPlaceholderNodeId && "DEATH1");
```
REMOVE: For debugging CI.

tcarmelveilleux commented on src/credentials/FabricTable.cpp:
```diff
@@ -146,6 +148,7 @@ void FabricInfo::operator=(FabricInfo && other)
     Reset();
 
     mNodeId                  = other.mNodeId;
+    VerifyOrDie(mNodeId != kPlaceholderNodeId && "DEATH2");
```
REMOVE: For debugging CI.

tcarmelveilleux committed:
Fix hermeticity of TC-CGEN scripts in CI

- The scripts did not factory reset properly so they were
  impacted by lack of hermeticity, causing a failure of CGEN-2.9
  after unrelated master changes.
- The method to find "commissioner's fabric" before some fabric
  removal was wrong in CGEN-2.9. Fixed the method to use CurrentFabricIndex

tcarmelveilleux committed:
Add logging

restyled-commits committed:
Restyled by autopep8

tcarmelveilleux committed:
Revert "Restyled by clang-format"

This reverts commit 66dee3d11e79f541215adfa2993c90da4294039d.

tcarmelveilleux committed:
Revert "Add a verification against some node ID to diagnose CI"

This reverts commit b188e365acd26ee7183846376d2a6b0102c21a8d.

tcarmelveilleux committed:
Merge branch 'fix-cgen-2.9' into add-set-vid-verify-statement

tcarmelveilleux committed:
Use fix-cgen-2.9

tcarmelveilleux committed:
Update cgen

bzbarsky-apple commented on src/credentials/PersistentStorageOpCertStore.cpp:
```diff
@@ -319,6 +386,97 @@ CHIP_ERROR PersistentStorageOpCertStore::UpdateOpCertsForFabric(FabricIndex fabr
+    ReturnErrorOnFailure(BasicVidVerificationAssumptionsAreMet(fabricIndex));
+    VerifyOrReturnError(vvsc.empty() || vvsc.size() <= Credentials::kMaxCHIPCertLength, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // Can't try to set a VVSC if ICAC present.
```
The API did not document that this is the responsibility of the callee to check.

In general, it would be nice if the cluster implementation could do these checks instead of every opcert store backend having to do them...  Not sure how viable that is.

tcarmelveilleux replied:
Will do

bzbarsky-apple commented on src/credentials/PersistentStorageOpCertStore.cpp:
```diff
@@ -319,6 +386,97 @@ CHIP_ERROR PersistentStorageOpCertStore::UpdateOpCertsForFabric(FabricIndex fabr
+                                                                                 ByteSpan vidVerificationStatement)
+{
+    ReturnErrorOnFailure(BasicVidVerificationAssumptionsAreMet(fabricIndex));
+    VerifyOrReturnError(vidVerificationStatement.empty() ||
```
Again, it would be nice if this size check were just in the cluster...

tcarmelveilleux replied:
I can do that for sure. I was trying to cover more unit tests in the opcerstore since the cluster is not unit testable yet, due to legacy design, but it has an integration test that may suffice.

bzbarsky-apple commented on src/lib/support/DefaultStorageKeyAllocator.h:
```diff
@@ -99,6 +99,11 @@ class DefaultStorageKeyAllocator
     static StorageKeyName FabricNOC(FabricIndex fabric) { return StorageKeyName::Formatted("f/%x/n", fabric); }
     static StorageKeyName FabricICAC(FabricIndex fabric) { return StorageKeyName::Formatted("f/%x/i", fabric); }
     static StorageKeyName FabricRCAC(FabricIndex fabric) { return StorageKeyName::Formatted("f/%x/r", fabric); }
+    static StorageKeyName FabricVVSC(FabricIndex fabric) { return StorageKeyName::Formatted("f/%x/vvvc", fabric); }
```
Why vvvc, not vvsc?  Worth documenting if there is a good reason.

tcarmelveilleux replied:
It's a typo, will fix.
